dict/dict.c

```diff
@@ -0,0 +1,270 @@
+  new_wc_ptr->next = word_offset;
+  new_wc_ptr->len = len;
+  ht_ptr->hash_bucket[index] = (unsigned long)ht_ptr->offset;
+  new_wc_ptr->word_offset = (unsigned long)CALC_ADDR(ht_ptr->offset, sizeof(WordChain));
+  strcpy((char*)CALC_ADDR(ht_ptr , new_wc_ptr->word_offset),str);
+
+  // Memory Management section 
+  ht_ptr->word_count++;
+  ht_ptr->offset = CALC_ADDR(ht_ptr->offset , CALC_ADDR(sizeof(WordChain) , len+1));
+  ht_ptr->used_bytes = ht_ptr->used_bytes + CALC_ADDR(sizeof(WordChain) , len+1);
+  ht_ptr->free_bytes = ht_ptr->free_bytes - CALC_ADDR(sizeof(WordChain) , len+1); 
+  printf("New Word = %s, len = %d, indexed = %d \n", 
+      CALC_ADDR(ht_ptr,new_wc_ptr->word_offset), new_wc_ptr->len, index);
+}
+
+char* search(HashTable *ht_ptr, char *str){
+  unsigned long index = index_of_string(str);
+  unsigned long key_len = strlen(str);
+  WordChain *wc_ptr = (ht_ptr->hash_bucket[index] == 0? NULL: 
+      (WordChain*)CALC_ADDR(ht_ptr , ht_ptr->hash_bucket[index]));
+  WordChain *temp = wc_ptr;
+
+  while(temp != NULL){
+    if(key_len == wc_ptr->len){
+
+      if(strcmp((char*)CALC_ADDR(ht_ptr , temp->word_offset),str) == 0) 
+        return str;
+      else 
+        temp = temp->next == 0? NULL: (WordChain*)CALC_ADDR(ht_ptr , temp->next);
+    }else
+      temp = temp->next == 0? NULL: (WordChain*)CALC_ADDR(ht_ptr , temp->next);
+  }
+
+  return NULL;
+}
+
+char* delete(HashTable *ht_ptr, char *str){
+  unsigned long index = index_of_string(str);
+  unsigned long key_len = strlen(str);
+  WordChain *wc_ptr = (ht_ptr->hash_bucket[index] == 0? NULL: 
+      (WordChain*)CALC_ADDR(ht_ptr , ht_ptr->hash_bucket[index]));
+  WordChain *temp = wc_ptr;
+  void *prev = 0;
+  void *holder = 0;
+  WordChain *temp_prev = 0;
+
+  while(temp != NULL){
+    if(key_len == wc_ptr->len){
+      if(strcmp((char*)CALC_ADDR(ht_ptr , temp->word_offset),str) == 0){
+        holder = (void*)ht_ptr->wc_fragment_chain_offset;
+        ht_ptr->wc_fragment_chain_offset = (unsigned long)temp - (unsigned long)ht_ptr;
+        if(prev == 0){
+          ht_ptr->hash_bucket[index] = temp->next;
+        }else{
+          temp_prev = (WordChain*)CALC_ADDR(ht_ptr,prev);
+          temp_prev->next = temp->next;
+        }
+        temp->next = (unsigned long)holder;
+        ht_ptr->fragment_count++;
+        ht_ptr->word_count--;
+        return str;
+      }
+      else{
+        prev = temp != NULL?(void*)((unsigned long)temp - (unsigned long)ht_ptr):NULL;
+        temp = temp->next == 0? NULL: (WordChain*)CALC_ADDR(ht_ptr , temp->next);
+      }
+    }else{
+      prev = temp != NULL?(void*)((unsigned long)temp - (unsigned long)ht_ptr):NULL;
+      temp = temp->next == 0? NULL: (WordChain*)CALC_ADDR(ht_ptr , temp->next);
+    }
+  }
+
+  return NULL;
+}
+
+
+int main(int arg_count, char **arg_str){
+  char *str;
+  int err;
+  HashTable *h_ptr = NULL;
+  err = init_hash_table(&h_ptr,MAX_SIZE_IN_BYTES);
```

Comment:
Will 4KB memory be enough to store 1M words?

What's the complexity of this algorithm? Hash table of size 100 storing 1M words will require 10K linked list traversals and comparisons to find/insert or delete a word.

Can you think of a more efficient solution?